osquery/tables/system/darwin/preferences.cpp

```diff
@@ -0,0 +1,179 @@
+
+#include <osquery/core.h>
+#include <osquery/logger.h>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+
+namespace osquery {
+namespace tables {
+
+// Walk the supported preference value structures.
+void genOSXPreferenceValues(const CFTypeRef& value,
+                            const Row& base,
+                            QueryData& results);
+
+void genOSXHashPreferences(const CFDictionaryRef& hash,
+                           const Row& base,
+                           QueryData& results) {
+  CFIndex count = CFDictionaryGetCount(hash);
+
+  // Allocate memory for copies of each key/value in dictionary.
+  auto keys = (void**)malloc(sizeof(void*) * count);
+  auto values = (void**)malloc(sizeof(void*) * count);
+  CFDictionaryGetKeysAndValues(hash, (const void**)keys, (const void**)values);
+
+  // Iterate over a preference value that contains a hash structure.
+  for (CFIndex i = 0; i < count; ++i) {
+    Row r = base;
+    if (CFGetTypeID((CFTypeRef)keys[i]) == CFStringGetTypeID()) {
+      if (r["subkey"].size() > 0) {
+        r["subkey"] += "/";
+      }
+      // The subkey is the hash map key.
+      r["subkey"] += stringFromCFString((CFStringRef)keys[i]);
+    }
+    genOSXPreferenceValues((CFTypeRef)values[i], r, results);
+  }
+
+  free(keys);
+  free(values);
+}
+
+void genOSXListPreferences(const CFArrayRef& list,
+                           const Row& base,
+                           QueryData& results) {
+  // Iterate over a preference value that contains a list structure.
+  for (CFIndex j = 0; j < CFArrayGetCount(list); ++j) {
+    Row r = base;
+
+    CFTypeRef value = (CFTypeRef)CFArrayGetValueAtIndex(list, j);
+    if (r["subkey"].size() > 0) {
+      r["subkey"] += "/";
+    }
+    // The subkey is the index into the list.
+    r["subkey"] += std::to_string(j);
+    genOSXPreferenceValues(value, r, results);
+  }
+}
+
+void genOSXPreferenceValues(const CFTypeRef& value,
+                            const Row& base,
+                            QueryData& results) {
+  if (value == nullptr) {
+    return;
+  }
+
+  // Emit a string representation for each preference type.
+  Row r = base;
+  if (CFGetTypeID(value) == CFNumberGetTypeID()) {
+    r["value"] = stringFromCFNumber((CFDataRef)value);
+  } else if (CFGetTypeID(value) == CFStringGetTypeID()) {
+    r["value"] = stringFromCFString((CFStringRef)value);
+  } else if (CFGetTypeID(value) == CFBooleanGetTypeID()) {
+    r["value"] =
+        (CFBooleanGetValue((CFBooleanRef)value) == TRUE) ? "true" : "false";
+  } else if (CFGetTypeID(value) == CFDataGetTypeID()) {
+    // Do not include data preferences.
+  } else if (CFGetTypeID(value) == CFArrayGetTypeID()) {
+    genOSXListPreferences((CFArrayRef)value, base, results);
+    return;
+  } else if (CFGetTypeID(value) == CFDictionaryGetTypeID()) {
+    // Generate a row for each hash key.
+    genOSXHashPreferences((CFDictionaryRef)value, base, results);
```

Comment:
A recursion limit would be nice, what do you think 10 or 20? After that I'd call foul, in practice the deepest I've seen is 5.